index.bs

```diff
@@ -5342,6 +5371,94 @@ over just using {{SyntaxError!!exception}} to refer to the {{DOMException}}. [[D
     </tbody>
 </table>
 
+<h4 id="idl-DOMException-derived-interfaces">{{DOMException}} derived interfaces</h4>
+
+When an exception needs to carry additional programmatically-introspectable information, beyond just
+its [=error name=], specification authors can create an [=interface=] which [=interface/inherits=]
+from {{DOMException}}. Such interfaces need to follow certain rules, in order to have a predictable
+shape for developers. Specifically:
+
+*   The [=identifier=] of the [=interface=] must end with <code>Error</code>, and must not be any
+    of the names in the [=error names table=].
+*   The [=interface=] must have a [=constructor operation=] which sets the instance's
+    [=DOMException/name=] to the interface's [=identifier=].
+*   Their [=constructor operation=] must take as its first parameter an [=optional=] {{DOMString}}
+    named |message| defaulting to the empty string, and must set the instance's
+    [=DOMException/message=] to |message|.
+*   Their [=constructor operation=] should take as its second parameter a [=dictionary=] containing
+    the additional information that needs to be exposed.
+*   They should have [=read only=] [=attributes=], whose names are the same as the members of the
+    constructor dictionary, which return the values accepted by the constructor operation.
+*   They should be [=serializable objects=], whose [=serialization steps=] and
+    [=deserialization steps=] preserve the additional information.
+
+<p class=note>These requirements mean that the inherited {{DOMException/code}} property of these
+interfaces will always return 0.
+
+<div class=example id=example-domexception-derived-interface>
+    The definition for a {{DOMException}} derived interface which carries along an additional
+    "hardware error code", for interfacing with a hypothetical Widget hardware device, could look
+    something like this:
+
+    <pre highlight=webidl>
+        [Exposed=Window]
+        interface WidgetError : DOMException {
+          constructor(optional DOMString message = "", WidgetErrorOptions options);
+
+          readonly attribute unsigned long long hardwareErrorCode;
```

Comment:
Perhaps we ought to encourage an enum instead?

Reply:
I'm kind of envisioning this being some opaque number spit out by the widget hardware. I can see how maybe that's not a good pattern to have as our only example, as we'd arguably want to have a web platform API abstract over that sort of thing (like the SmartCard proposal is doing).

Using an enum, to me, gets into #1219, which I don't consider quite settled yet.

We should probably change this to be some number spit out by a protocol, instead of by a widget, since that's a bit more respectable and has precedent in both RTCError and WebTransportError.